Library/Homebrew/cask/tab.rb

```diff
@@ -0,0 +1,184 @@
+    #
+    # NOTE: Results are cached.
+    sig { params(path: Pathname).returns(Tab) }
+    def self.from_file(path)
+      cache.fetch(path) do |p|
+        content = File.read(p)
+        return empty if content.blank?
+
+        cache[p] = from_file_content(content, p)
+      end
+    end
+
+    # Like {from_file}, but bypass the cache.
+    sig { params(content: String, path: Pathname).returns(Tab) }
+    def self.from_file_content(content, path)
+      attributes = begin
+        JSON.parse(content)
+      rescue JSON::ParserError => e
+        raise e, "Cannot parse #{path}: #{e}", e.backtrace
+      end
+      attributes["tabfile"] = path
+
+      new(attributes)
+    end
+
+    sig { params(cask: Cask).returns(Tab) }
+    def self.for_cask(cask)
+      path = cask.metadata_main_container_path/FILENAME
+
+      return from_file(path) if path.exist?
+
+      tab = empty
+      tab.source = {
+        "path"         => cask.sourcefile_path.to_s,
+        "tap"          => cask.tap&.name,
+        "tap_git_head" => nil,
+        "version"      => cask.version.to_s,
+      }
+      tab.artifacts = cask.to_h["artifacts"]
+      tab.source["tap_git_head"] = cask.tap.git_head if cask.tap&.installed?
+
+      tab
+    end
+
+    sig { returns(Tab) }
+    def self.empty
+      attributes = {
+        "homebrew_version"        => HOMEBREW_VERSION,
+        "loaded_from_api"         => false,
+        "installed_as_dependency" => false,
+        "installed_on_request"    => false,
+        "time"                    => nil,
+        "dependencies"            => nil,
+        "arch"                    => nil,
+        "source"                  => {
+          "path"         => nil,
+          "tap"          => nil,
+          "tap_git_head" => nil,
+          "version"      => nil,
+        },
+        "installed_on"            => DevelopmentTools.generic_build_system_info,
+        "artifacts"               => [],
+      }
+
+      new(attributes)
+    end
+
+    def self.runtime_deps_hash(cask, depends_on)
+      mappable_types = [:cask, :formula]
+      depends_on.to_h do |type, deps|
+        next [type, deps] unless mappable_types.include? type
+
+        deps = deps.map do |dep|
+          if type == :cask
+            c = CaskLoader.load(dep)
+            {
+              "full_name"         => c.full_name,
+              "version"           => c.version.to_s,
+              "declared_directly" => cask.depends_on.cask.include?(dep),
+            }
+          elsif type == :formula
+            f = Formulary.factory(dep, warn: false)
+            {
+              "full_name"         => f.full_name,
+              "version"           => f.version.to_s,
+              "revision"          => f.revision,
+              "pkg_version"       => f.pkg_version.to_s,
+              "declared_directly" => cask.depends_on.formula.include?(dep),
+            }
+          else
+            dep
+          end
```

Comment:
Question: What types of deps are we expecting here that are not casks or formulae?

Reply:
`:macos` or `:arch`